cluster-autoscaler/proposals/plugable-provider-grpc.md

```diff
@@ -0,0 +1,127 @@
+
+There are two alternatives on how to plug custom cloud providers:
+
+* **Option1:** Install the plugin as a binary that would be mounted into the CA container
+and invoked by CA server.
+In this option CA server launches each provider plugin as a subprocess and communicates with it over gRPC.
+
+* **Option2:** A custom cloud provider server is deployed along side CA and both communicates via gRPC with TLS/SSL.
+
+Regardless of the chosen option, both solutions have to expose a common gRPC API
+with the following operations:
+
+```go
+type clusterAutoscalerCustomProviderServer struct {
+        ...
+}
+
+func (s *clusterAutoscalerCustomProviderServer) NodeGroups(ctx context.Context) ([]pb.NodeGroup, error) {
+        ...
+}
+...
+
+func (s *clusterAutoscalerCustomProviderServer) NodeGroupForNode(ctx context.Context, *apiv1.Node) (*pb.NodeGroup, error) {
+        ...
+}
+...
+
+func (s *clusterAutoscalerCustomProviderServer) Refresh() error {
+        ...
+}
+...
+
+func (s *clusterAutoscalerCustomProviderServer) GetAvailableMachineTypes() ([]string, error) {
+        ...
+}
+
+...
+
+```
+
+Obviously, these API calls implement the `CloudProvider` interface of the [CA](https://github.com/kubernetes/autoscaler/blob/master/cluster-autoscaler/cloudprovider/cloud_provider.go#L50):
+
+```go
+type CloudProvider interface {
+	Name() string
+
+	NodeGroups() []NodeGroup
+
+	NodeGroupForNode(*apiv1.Node) (NodeGroup, error)
+
+	Pricing() (PricingModel, errors.AutoscalerError)
+
+	GetAvailableMachineTypes() ([]string, error)
+
+	NewNodeGroup(machineType string, labels map[string]string, systemLabels map[string]string,
+		taints []apiv1.Taint, extraResources map[string]resource.Quantity) (NodeGroup, error)
+
+	GetResourceLimiter() (*ResourceLimiter, error)
+
+	GPULabel() string
+
+	GetAvailableGPUTypes() map[string]struct{}
+
+	Cleanup() error
+
+	Refresh() error
+
+  ...
+}
+```
+
+In order to talk to the custom cloud provider server, this new cloud provider has to be registered
+when bootstrapping the CA. 
+Consequently, the CA needs to expose new flags to specify the cloud provider and all the required properties
+to reach the remote gRPC server.
+
+A new flag, named
+`--cloud-provider-url=https://local.svc.io/mycloudprovider/server`, determines the URL to reach the custom provider implementation.
+In addition this approach reuses the existing flag that defines the name of the cloud provider `--cloud-provider=mycloudprovider`
+https://github.com/kubernetes/autoscaler/tree/master/cluster-autoscaler/cloudprovider.
+
+To connect the CA core with this new remote cloud provider, this approach needs to implement a new generic cloud provider 
+as part of the CA core code.
+This new provider, named `CustomCloudProvider` simply makes gRPC calls to the remote functions exposed by the custom cloud provider server. In other words, it forwards the calls and handle the errors analogously how done in other existing providers.
```

Comment:
imo, we should name this `gRPCCloudProvider` or something similar to be more explicit in the name.

Reply:
+1 for custom being too general. ExternalCloudProvider maybe? Naming stuff is hard :)

Reply:
I'd propose `http`, otherwise I'd go with `external`

Reply:
May be ExternalgRPCCloudProvider?